Update cookiecutter template versioning scheme

HealthyPear opened the pull request:
This is an attempt to make template versioning a bit more dynamic and robust.

This is a summary of changes:

- introduced a new flag option to `showyourwork setup` called `--action-version` which specifies the version of the GitHub action (complementary to `--action-spec` which instead specifies the version of showyourwork that the action should use - I would even change the name of the latter in "showyourwork-remote-version" to make it clearer)
- change the default behavior of `--action-spec` to use on the remote the same version of showyourwork that the user is using to create the new project
- set the default behavior of `--action-version` to check if there are new merged PR into the main version (in this case use the latest commit) and if not use the latest release (for now v1)
- add new unit tests to make sure the cookiecutter templates contain valid YAML and have the expected verions.

On top of all local and unit tests passing I also created a [test repository](https://github.com/HealthyPear/test_pr_576) for this PR in place of the remote tests that are currently out of order to check both the build workflows and build PR workflows.

While I wait for reviews I will update the docs and changelog accordingly.

Closes #507.

HealthyPear committed:
Update .gitignore

Ignoring now cache from mypy, pytest and ruff

HealthyPear committed:
Improve version handling in setup command

- add action_version parameter to setup command and add it to cookecutter template
- default action_spec to local showyourwork version if not provided
- default action_version to latest release (or latest dev commit if newer) if not provided

Set action_spec to match local version if not provided

Now checks git status first, then combines with version analysis

Remove use of capture_output get_stdout

HealthyPear committed:
Update cookiecutter build PR workflow for new versioning scheme

re-introduce action_spec in build PR template

HealthyPear committed:
Update cookiecutter build workflow for new versioning scheme

re-introduce action_spec in build template

HealthyPear committed:
Update cookiecutter process PR workflow for new versioning scheme

HealthyPear committed:
Add unit tests for workflow templates

HealthyPear committed:
Add news fragment for PR #576

arm61 commented:
@HealthyPear I can review this if you would like?

HealthyPear commented:
> @HealthyPear I can review this if you would like?

Hi! Sure thank you I did not want to bother since I saw you were not active on this repo for a while. I will accept all the help of course!

arm61 reviewed (approved):
I have had a good read through the code and everything seems sensible to me. I have *not* built and tested on a test repo, but @HealthyPear's test repo look logical to me.

HealthyPear commented on src/showyourwork/cli/commands/setup.py:
```diff
@@ -47,6 +50,59 @@ def callback(code, stdout, stderr):
+
+    # Set action_spec to local showyourwork version if not provided
+    if not action_spec:
+        action_spec = __version__
```
Actually I see an issue here which might be very wrong

this should not be the version but something like 

`git+https://github.com/showyourwork/showyourwork`

so right now this will make any project remote fail

arm61 replied:
Ah, I didn't spot that. How did your test repo pass?

HealthyPear replied:
because since I had to test this specific branch (of the PR) when I created the project I did it passing the spec with the `-a` option as

`showyourwork setup -a git+https://github.com/showyourwork/showyourwork@fix-cookiecutter_template_versions foo/bar`

and so this conditional statement never went through

(which is actually a missing remote test)

before I add the test I need to figure out if we are covering all the cases: the obvious solution is to set 

`git+https://github.com/showyourwork/showyourwork`

am I missing something?

arm61 replied:
It makes sense to me... I think

HealthyPear replied:
OK I fixed it, see below in the thread for a detailed explanation

HealthyPear committed:
Add unit tests for action_spec resolution logic

Make tests class based and add test cases

HealthyPear commented:
OK now it should be correct and way more robust.

Basically the issue was that I was mistakenly using the _version_ of the package to set the action spec in the cookiecutter template, but what you want is something that can be fed to `pip install`.

The basic problem that triggered this PR is that there is no new release since 2+ years (see #563) so many people was installing locally the _development_ version of showyourwork, but the default behavior of the `--action-spec` flag (and thus of the action itself) was to use the latest release on PyPI on the remote which of course made things crash on the remote and people had to manually update their github workflows every time they started a new project.

The behavior now for action spec is to use the version which is locally installed at the time of project creation.

This opened a kind of a pandora box, because the amount of cases is not just release vs dev version!

Also using the version is not the most robust thing to do in the case of a development version because the version file is created at the time of installation and cached after that!

So this PR now uses directly git to asses the status of your local working tree and comparing it to what your branch (any branch!) is tracking on the remote (of any!):

- **Uncommitted changes**: Local working tree has unstaged/staged changes
- **Unpushed commits**: Local branch is ahead of remote tracking branch
- **Both uncommitted and unpushed**: Combination of both scenarios
- **No tracking branch**: Branch exists locally but has no remote tracking branch
- **Development versions**: Versions with .dev, +, .dirty, or rc suffixes

For each case, you will get the "nearest" development version on the remote and a warning will be logged in case some differences have been detected.

For example in my current case my installed version says `0.4.4.dev219+gf6a876582`, meaning that I am 219 commits after v0.4.3 and I ran the installation when I was at commit `f6a876582`.
But the reality is that my git log shows (after rebasing and force-pushing)

```shell
4020b56 (HEAD -> fix-cookiecutter_template_versions, upstream/fix-cookiecutter_template_versions) Add unit tests for action_spec resolution logic
96eaf84 Add news fragment for PR #576
216d433 Add unit tests for workflow templates
b36c151 Update cookiecutter process PR workflow for new versioning scheme
52f50b5 Update cookiecutter build workflow for new versioning scheme
2d1cd76 Update cookiecutter build PR workflow for new versioning scheme
89daea9 Improve version handling in setup command
24afac8 Update .gitignore
```

so not only I am way after that commit, but it does not even exist anymore!

So now if in this case I have pushed all my local commits on this branch, if I do `showyourwork setup foo/bar` the version used by showyourwork (if I do not specify a specific `--action-spec` of course) will be

```yml
uses: showyourwork/showyourwork-action@085c01791d75285222ba2345b93184046f6e9672
        with:
          showyourwork-spec: git+https://github.com/showyourwork/showyourwork@fix-cookiecutter_template_versions
```

meaning use the latest dev version of the action (because it detected commits on the main branch after the latest release) and the latest commit on this very branch because I pushed everything - if this was not the case it would do the same but printing a warning.

😮‍💨

HealthyPear committed:
Update setup command

- make sure git commands are run where showyourwork is installed
- if not a git repo catch specific expection and pass
- minor formatting

HealthyPear commented:
Just made a last update:

- the git commands where run in the cwd but of course this is wrong, they should run only where showyourwork is installed!
- in case showyourwork is installed from a release, that would not be a git repo so the command will catch a specific exception (`exceptions.CalledProcessError`) before passing - thereby using the locally installed version - instead of a generic `Exception`

HealthyPear committed:
Add admonition boxes about versions in Known Issues page
